File: go/sdk/alarm_contract_test.go
package sdk

import (
	"encoding/json"
	vdmmqttv1 "github.com/inteagle-vision/inteagle-vdm-mqtt-examples/go/generated"
	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/proto"
	"os"
	"testing"
)

func alarmExamples(t *testing.T, name string) []map[string]json.RawMessage {
	t.Helper()
	raw, err := os.ReadFile("../../examples/alarms/" + name + ".json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []map[string]json.RawMessage
	if err := json.Unmarshal(raw, &cases); err != nil {
		t.Fatal(err)
	}
	return cases
}

func TestAlarmConfigurationExamples(t *testing.T) {
	for _, item := range alarmExamples(t, "requests") {
		var method string
		if err := json.Unmarshal(item["method"], &method); err != nil {
			t.Fatal(err)
		}
		for _, format := range []PayloadFormat{Protobuf, JSON} {
			var params map[string]any
			if err := json.Unmarshal(item[string(format)], &params); err != nil {
				t.Fatal(err)
			}
			raw, _, err := (Codec{Format: format}).EncodeRPC(method, params, 91)
			if err != nil {
				t.Fatalf("%s: %v", item["name"], err)
			}
			if format == Protobuf {
				wantBytes, err := json.Marshal(map[string]any{"schemaVersion": 1, "reqId": 91, method: params})
				if err != nil {
					t.Fatal(err)
				}
				want, got := &vdmmqttv1.RpcRequest{}, &vdmmqttv1.RpcRequest{}
				if err := protojson.Unmarshal(wantBytes, want); err != nil {
					t.Fatal(err)
				}
				if err := proto.Unmarshal(raw, got); err != nil {
					t.Fatal(err)
				}
				if !proto.Equal(want, got) {
					t.Fatalf("wrong %s request: %v", method, got)
				}
			} else {
				var got map[string]json.RawMessage
				if err := json.Unmarshal(raw, &got); err != nil {
					t.Fatal(err)
				}
				var actual map[string]any
				if err := json.Unmarshal(got["params"], &actual); err != nil {
					t.Fatal(err)
				}
				a, _ := json.Marshal(actual)
				b, _ := json.Marshal(params)
				if string(a) != string(b) {
					t.Fatalf("JSON params changed: %s", raw)
				}
			}
		}
	}
}

func TestAlarmLifecycleExamples(t *testing.T) {
	topics, _ := TopicsForDevice("DEMO001")
	for _, item := range alarmExamples(t, "events") {
		for _, format := range []PayloadFormat{Protobuf, JSON} {
			raw := []byte(item[string(format)])
			var expected map[string]any
			if err := json.Unmarshal(raw, &expected); err != nil {
				t.Fatal(err)
			}
			if format == Protobuf {
				message := &vdmmqttv1.Alarm{}
				if err := protojson.Unmarshal(raw, message); err != nil {
					t.Fatal(err)
				}
				var err error
				raw, err = proto.Marshal(message)
				if err != nil {
					t.Fatal(err)
				}
			}
			decoded, err := (Codec{Format: format}).Decode(topics.Topic("3A"), topics, raw)
			if err != nil {
				t.Fatal(err)
			}
			fields, err := decoded.AsMap()
			if err != nil {
				t.Fatal(err)
			}
			_, hasLevel := fields["level"]
			_, wantLevel := expected["level"]
			if fields["eventId"] != expected["eventId"] || fields["alarmId"] != expected["alarmId"] || hasLevel != wantLevel {
				t.Fatalf("lost ID precision or optional level: %v", fields)
			}
		}
	}
}

File: java/src/test/java/com/inteagle/vdm/mqtt/sdk/AlarmContractTest.java
package com.inteagle.vdm.mqtt.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.inteagle.vdm.mqtt.v1.Alarm;
import com.inteagle.vdm.mqtt.v1.RpcRequest;
import com.google.protobuf.util.JsonFormat;
import java.nio.file.Path;
import java.util.Map;
import org.junit.jupiter.api.Test;

final class AlarmContractTest {
  private final ObjectMapper mapper = new ObjectMapper();
  @Test void configurationExamples() throws Exception {
    var cases = mapper.readTree(Path.of("../examples/alarms/requests.json").toFile());
    for (var item : cases) {
      for (var format : PayloadFormat.values()) {
        Map<String, Object> params = mapper.convertValue(item.get(format.toString()), new TypeReference<>() {});
        String method = item.get("method").asText();
        var encoded = new VdmCodec(format).encodeRpcRequest(method, params, 91);
        if (format == PayloadFormat.PROTOBUF) {
          ObjectNode want = mapper.createObjectNode().put("schemaVersion",1).put("reqId",91);
          want.set(method,item.get("protobuf"));
          var builder = RpcRequest.newBuilder();
          JsonFormat.parser().merge(want.toString(),builder);
          assertEquals(builder.build(),RpcRequest.parseFrom(encoded.payload()),item.get("name").asText());
        } else {
          assertEquals(item.get("json"),mapper.readTree(encoded.payload()).get("params"));
        }
      }
    }
  }
  @Test void lifecycleExamples() throws Exception {
    var topics = VdmTopics.forDevice("DEMO001");
    var cases = mapper.readTree(Path.of("../examples/alarms/events.json").toFile());
    for (var item : cases) {
      for (var format : PayloadFormat.values()) {
        var value = item.get(format.toString());
        byte[] raw = mapper.writeValueAsBytes(value);
        if (format == PayloadFormat.PROTOBUF) {
          var builder = Alarm.newBuilder();
          JsonFormat.parser().merge(value.toString(),builder);
          raw = builder.build().toByteArray();
        }
        var decoded = new VdmCodec(format).decode(topics.topic("3A"),topics,raw);
        assertEquals(value.get("eventId"),decoded.data().get("eventId"));
        assertEquals(value.get("alarmId"),decoded.data().get("alarmId"));
        assertEquals(value.has("level"),decoded.data().has("level"));
        if (decoded.value() instanceof Alarm alarm) assertEquals(value.has("level"),alarm.hasLevel());
      }
    }
  }
}
